frontend/packages/console-plugin-sdk/src/codegen/index.test.ts

```diff
@@ -0,0 +1,104 @@
+import {
```

Comment:
Why aren't we following the `module.spec.ts` filename pattern here?

Reply:
You're right :+1: we should follow the existing pattern.

I'll do the change and update Jest `testRegex` value accordingly.

@alecmerdler @spadgett What's your opinion on unit test co-location? Right now, the test sits next to the corresponding module. An alternative is to have e.g.

```
packages/console-plugin-sdk/__tests__/codegen/index.spec.ts
```

which adds structural duplicity in favor of better test isolation.

Once we move code out of the `public` directory into appropriate packages, which test location convention should we follow?

Reply:
I prefer having tests next to their source file, but that's not what we had when I started working on `bridge`, so we've stuck with the `__tests__` directory. We should stick with the duplicate structure; following the `module.spec.ts` filename format makes it easy to find the test files using search anyway.

Reply:
Why not just follow the typical pattern found and described by most react apps and tutorials wherein the `__tests__` directory is located adjacent to the file being tested. So in this case create test file at `src/codegen/__tests__/index.spec.ts`

Reply:
> I prefer having tests next to their source file

I have the same preference, the main reason being that it's immediately clear if the given module has a corresponding test, and the secondary reason being that it encourages developers to treat tests in the same way as actual code.

I'm OK with @christianvogt's proposal, co-locating tests under `__tests__` adjacent to the module being tested. Should I proceed with this change? @spadgett @alecmerdler 

The underscores in `__tests__` is something I don't like, though. I guess this comes from Python, which uses `__` prefix to denote things that are outside of the regular code. (From my point of view, unit tests should be an integral part of the regular code, not something outside of it.)

Reply:
If we are going to co-locate the tests, my vote would to have a flat structure and not use `__tests__`. So: 
```
components/
|--- index.ts
|--- index.spec.ts
|--- widget.tsx
|--- widget.spec.tsx
```
Whatever we choose, we should do it across the entire codebase to ensure consistency, which means breaking out the existing `__tests__` directory to their respective source directories.

Reply:
> Whatever we choose, we should do it across the entire codebase to ensure consistency

I'll take care of that as part of the "moving stuff out of `public` directory" effort.

Reply:
I agree that we should do it across the whole repo. But do so in steps.
One benefit to the `__tests__` dir is that related test files can also be placed in there as to not confused with application code. The double underscores help move the dir to the top of the file sort order. I can appreciate having tests as sibling files though.

When using jest we also have a `__mocks__` folder for mocks and a `__snapshots__` folder for snapshots. Snapshots could probably be co-located once again in the same dir. TBH I never looked at changing the directory for mocks though.

Something worth bringing up as we look to improve and enhance the coding guidelines as well as look to add more linting.

Reply:
> One benefit to the `__tests__` dir is that related test files can also be placed in there as to not confused with application code. The double underscores help move the dir to the top of the file sort order. I can appreciate having tests as sibling files though.

These are good points. @spadgett do you have any thoughts on this?

For this PR, I'll use the adjacent `__tests__` directory convention:

```
path/to/module.ts(x)
path/to/__tests__/module.spec.ts(x)
```

> When using jest we also have a `__mocks__` folder for mocks and a `__snapshots__` folder for snapshots.

Actually, some files under the `__mocks__` directory are [setup files](https://jestjs.io/docs/en/configuration.html#setupfiles-array):

```
"setupFiles": [
  "./__mocks__/requestAnimationFrame.js",
  "./__mocks__/localStorage.ts",
  "./__mocks__/matchMedia.js",
  "./before-tests.js"
],
```

Some other files in that directory are [automatic module mocks](https://jestjs.io/docs/en/configuration.html#modulenamemapper-object-string-string):

```
"moduleNameMapper": {
  "\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "<rootDir>/__mocks__/fileMock.js",
  "\\.(css|less)$": "<rootDir>/__mocks__/styleMock.js"
},
```

The remaining files in that directory are what I call helper mocks. These are used in tests to provide mocked versions of specific objects, for example:

```
// in __tests__/components/catalog.spec.tsx
import { catalogListPageProps, catalogItems, catalogCategories } from '../../__mocks__/catalogItemsMocks';
```

As far as I can see, Console doesn't use Jest [manual mocks](https://jestjs.io/docs/en/manual-mocks) which are used to stub out entire modules, both application and vendor.

> Snapshots could probably be co-located once again in the same dir. TBH I never looked at changing the directory for mocks though.

Jest originally introduced the `__snapshots__` directory for the purpose of completely unambiguous test vs. snapshot file resolution.

As of Jest 24, it's [possible to customize the test snapshot location](https://github.com/facebook/jest/issues/1650#issuecomment-435545153) by providing a custom [snapshot resolver](https://jestjs.io/docs/en/configuration.html#snapshotresolver-string).

I don't see any value in having many `__snapshots__` directories dispersed across the codebase, so I'd vote for co-locating test snapshot right next to the test itself. If we go with the adjacent `__tests__` convention, this means:

```
path/to/module.ts(x)
path/to/__tests__/module.spec.ts(x)
path/to/__tests__/module.spec.snap
```

Reply:
Also, by putting test snapshots next to the corresponding tests,

> related test files can also be placed in there as to not confused with application code

makes even more sense.